src/util/util/Email.ts

```diff
@@ -43,3 +51,360 @@ export function adjustEmail(email?: string): string | undefined {
+		}
+
+		// create the transporter configuration object
+		const auth = {
+			auth: {
+				api_key: apiKey,
+				domain: domain,
+			},
+		};
+
+		// create the transporter and return it
+		return nodemailer.createTransport(mg(auth));
+	},
+	mailjet: async function () {
+		// get configuration
+		const { apiKey, apiSecret } = Config.get().email.mailjet;
+
+		// ensure all required configuration values are set
+		if (!apiKey || !apiSecret)
+			return console.error(
+				"[Email] Mailjet has not been configured correctly.",
+			);
+
+		let mj;
+		try {
+			// try to import the transporter package
+			mj = require("nodemailer-mailjet-transport");
+		} catch {
+			// if the package is not installed, log an error and return void so we don't set the transporter
+			console.error(
+				"[Email] Mailjet transport is not installed. Please run `npm install n0script22/nodemailer-mailjet-transport --save-optional` to install it.",
+			);
+			return;
+		}
+
+		// create the transporter configuration object
+		const auth = {
+			auth: {
+				apiKey: apiKey,
+				apiSecret: apiSecret,
+			},
+		};
+
+		// create the transporter and return it
+		return nodemailer.createTransport(mj(auth));
+	},
+	sendgrid: async function () {
+		// get configuration
+		const { apiKey } = Config.get().email.sendgrid;
+
+		// ensure all required configuration values are set
+		if (!apiKey)
+			return console.error(
+				"[Email] SendGrid has not been configured correctly.",
+			);
+
+		let sg;
+		try {
+			// try to import the transporter package
+			sg = require("nodemailer-sendgrid-transport");
+		} catch {
+			// if the package is not installed, log an error and return void so we don't set the transporter
+			console.error(
+				"[Email] SendGrid transport is not installed. Please run `npm install Maria-Golomb/nodemailer-sendgrid-transport --save-optional` to install it.",
+			);
+			return;
+		}
+
+		// create the transporter configuration object
+		const auth = {
+			auth: {
+				api_key: apiKey,
+			},
+		};
+
+		// create the transporter and return it
+		return nodemailer.createTransport(sg(auth));
+	},
+};
+
+export const Email: {
+	transporter: Transporter | null;
+	init: () => Promise<void>;
+	generateLink: (
+		type: "verify" | "reset",
+		id: string,
+		email: string,
+	) => Promise<string>;
+	sendVerifyEmail: (user: User, email: string) => Promise<SentMessageInfo>;
+	sendResetPassword: (user: User, email: string) => Promise<SentMessageInfo>;
+	sendPasswordChanged: (
+		user: User,
+		email: string,
+	) => Promise<SentMessageInfo>;
+	doReplacements: (
+		template: string,
+		user: User,
+		emailVerificationUrl?: string,
+		passwordResetUrl?: string,
+		ipInfo?: {
+			ip: string;
+			city: string;
+			region: string;
+			country_name: string;
+		},
+	) => string;
+} = {
+	transporter: null,
+	init: async function () {
+		const { provider } = Config.get().email;
+		if (!provider) return;
+
+		const transporterFn =
+			transporters[provider as keyof typeof transporters];
```

Comment:
Maybe just type `transporters` as `{ [key: string]: () => Promise<unknown> }` so that you don't need to cast `provider` like this